File: src/Atlasd/Battlenet/Protocols/Game/Messages/SID_FRIENDSLIST.cs
using Atlasd.Battlenet.Exceptions;
using Atlasd.Daemon;
using System;
using System.Collections.Generic;
using System.IO;
using System.Text;

namespace Atlasd.Battlenet.Protocols.Game.Messages
{
    class SID_FRIENDSLIST : Message
    {

        public SID_FRIENDSLIST()
        {
            Id = (byte)MessageIds.SID_FRIENDSLIST;
            Buffer = new byte[0];
        }

        public SID_FRIENDSLIST(byte[] buffer)
        {
            Id = (byte)MessageIds.SID_FRIENDSLIST;
            Buffer = buffer;
        }

        public override bool Invoke(MessageContext context)
        {
            switch (context.Direction)
            {
                case MessageDirection.ClientToServer:
                    {
                        Logging.WriteLine(Logging.LogLevel.Debug, Logging.LogType.Client_Game, context.Client.RemoteEndPoint, $"[{Common.DirectionToString(context.Direction)}] {MessageName(Id)} ({4 + Buffer.Length} bytes)");

                        if (Buffer.Length != 0)
                            throw new GameProtocolViolationException(context.Client, $"{MessageName(Id)} buffer must be 0 bytes");

                        if (context.Client.GameState == null || context.Client.GameState.ActiveAccount == null)
                            throw new GameProtocolViolationException(context.Client, $"{MessageName(Id)} cannot be processed without an active login");

                        return new SID_FRIENDSLIST().Invoke(new MessageContext(context.Client, MessageDirection.ServerToClient));
                    }
                case MessageDirection.ServerToClient:
                    {
                        /**
                         * (UINT8) Number of entries
                         *
                         * For each entry:
                         *   (STRING) Account name
                         *    (UINT8) Status
                         *    (UINT8) Location id
                         *    UINT32) Product id
                         *   (STRING) Location name
                         */

                        var bufferSize = (uint)1;
                        var friends = new List<Friend>();
                        var friendStrings = (List<byte[]>)context.Client.GameState.ActiveAccount.Get(Account.FriendsKey, new List<byte[]>());

                        foreach (var friendString in friendStrings)
                        {
                            var friend = new Friend(context.Client.GameState, friendString);
                            friends.Add(friend);
                            bufferSize += (uint)(8 + friend.Username.Length + friend.LocationString.Length);

                            if (friends.Count == 255) // Hard limit based on counter in message format
                            {
                                Logging.WriteLine(Logging.LogLevel.Warning, Logging.LogType.Client_Game, $"Hard limit of 255 friends reached, dropping remaining {friendStrings.Count - 255} friends from {MessageName(Id)} reply");
                                break;
                            }
                        }

                        Buffer = new byte[bufferSize];

                        using var m = new MemoryStream(Buffer);
                        using var w = new BinaryWriter(m);

                        w.Write((byte)friends.Count);
                        foreach (var friend in friends)
                        {
                            w.WriteByteString(friend.Username);
                            w.Write((byte)friend.StatusId);
                            w.Write((byte)friend.LocationId);
                            w.Write((uint)friend.ProductCode);
                            w.WriteByteString(friend.LocationString);
                        }

                        Logging.WriteLine(Logging.LogLevel.Debug, Logging.LogType.Client_Game, context.Client.RemoteEndPoint, $"[{Common.DirectionToString(context.Direction)}] {MessageName(Id)} ({4 + Buffer.Length} bytes)");
                        context.Client.Send(ToByteArray(context.Client.ProtocolType));
                        return true;
                    }
            }

            return false;
        }
    }
}

File: src/AtlasV/Battlenet/Protocols/Game/Messages/SID_FRIENDSLIST.vb
Imports AtlasV.Battlenet.Exceptions
Imports AtlasV.Daemon
Imports System
Imports System.Collections.Generic
Imports System.IO
Imports System.Text

Namespace AtlasV.Battlenet.Protocols.Game.Messages
    Class SID_FRIENDSLIST
        Inherits Message

        Public Sub New()
            Id = CByte(MessageIds.SID_FRIENDSLIST)
            Buffer = Array.Empty(Of Byte)()
        End Sub

        Public Sub New(ByVal varBuffer As Byte())
            Id = CByte(MessageIds.SID_FRIENDSLIST)
            Buffer = varBuffer
        End Sub

        Public Overrides Function Invoke(ByVal context As MessageContext) As Boolean
            Select Case context.Direction
                Case MessageDirection.ClientToServer
                    Logging.WriteLine(Logging.LogLevel.Debug, Logging.LogType.Client_Game, context.Client.RemoteEndPoint, $"[{Common.DirectionToString(context.Direction)}] {MessageName(Id)} ({4 + Buffer.Length} bytes)")
                    If Buffer.Length <> 0 Then Throw New GameProtocolViolationException(context.Client, $"{MessageName(Id)} buffer must be 0 bytes")
                    If context.Client.GameState Is Nothing OrElse context.Client.GameState.ActiveAccount Is Nothing Then Throw New GameProtocolViolationException(context.Client, $"{MessageName(Id)} cannot be processed without an active login")
                    Return New SID_FRIENDSLIST().Invoke(New MessageContext(context.Client, MessageDirection.ServerToClient))
                Case MessageDirection.ServerToClient
                    Dim bufferSize = CUInt(1)
                    Dim friends = New List(Of [Friend])()
                    Dim friendStrings = CType(context.Client.GameState.ActiveAccount.[Get](Account.FriendsKey, New List(Of Byte())(),), List(Of Byte()))

                    For Each friendString In friendStrings
                        Dim [friend] = New [Friend](context.Client.GameState, friendString)
                        friends.Add([friend])
                        bufferSize += CUInt((8 + [friend].Username.Length + [friend].LocationString.Length))

                        If friends.Count = 255 Then
                            Logging.WriteLine(Logging.LogLevel.Warning, Logging.LogType.Client_Game, $"Hard limit of 255 friends reached, dropping remaining {friendStrings.Count - 255} friends from {MessageName(Id)} reply")
                            Exit For
                        End If
                    Next

                    Buffer = New Byte(bufferSize - 1) {}

                    Using m = New MemoryStream(Buffer)
                        Using w = New BinaryWriter(m)
                            w.Write(CByte(friends.Count))

                            For Each [friend] In friends
                                w.WriteByteString([friend].Username)
                                w.Write(CByte([friend].StatusId))
                                w.Write(CByte([friend].LocationId))
                                w.Write(CUInt([friend].ProductCode))
                                w.WriteByteString([friend].LocationString)
                            Next
                        End Using
                    End Using

                    Logging.WriteLine(Logging.LogLevel.Debug, Logging.LogType.Client_Game, context.Client.RemoteEndPoint, $"[{Common.DirectionToString(context.Direction)}] {MessageName(Id)} ({4 + Buffer.Length} bytes)")
                    context.Client.Send(ToByteArray(context.Client.ProtocolType))
                    Return True
            End Select

            Return False
        End Function
    End Class
End Namespace
